docs/painless/painless-guide/painless-datetime.asciidoc

```diff
@@ -0,0 +1,321 @@
+
+===== Datetime Formatting Examples
+
+* format to a String (ISO 8601)
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+String datetime = zdt.format(DateTimeFormatter.ISO_INSTANT);
+----
+Note the specification of a built-in DateTimeFormatter.
++
+* format to a String (custom)
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+DateTimeFormatter dtf = DateTimeFormatter.ofPattern(
+        "'date:' yyyy/MM/dd 'time:' HH:mm:ss");
+String datetime = zdt.format(dtf);
+----
+Note the specification of a custom DateTimeFormatter.
+
+==== Datetime Conversion
+
+Datetime conversion is a switch from a long representation to a ZonedDateTime
+representation and vice versa.
+
+===== Datetime Conversion Examples
+
+* convert from milliseconds
++
+[source,Painless]
+----
+long milliSinceEpoch = 434931330000L;
+Instant instant = Instant.ofEpochMilli(milliSinceEpoch);
+ZonedDateTime zdt = ZonedDateTime.ofInstant(instant, ZoneId.of('Z'));
+----
++
+* convert to milliseconds
++
+[source,Painless]
+-----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+long milliSinceEpoch = zdt.toInstant().toEpochMilli();
+-----
+
+==== Datetime Pieces
+
+Use the ZonedDateTime
+<<painless-api-reference-shared-ZonedDateTime, method of>> to create a new
+ZonedDateTime from pieces (year, month, day, hour, minute, second, nano,
+time zone). Use ZonedDateTime
+<<painless-api-reference-shared-ZonedDateTime, methods>> to extract pieces from
+a ZonedDateTime.
+
+===== Datetime Pieces Examples
+
+* create a ZonedDateTime from pieces
++
+[source,Painless]
+----
+int year = 1983;
+int month = 10;
+int day = 13;
+int hour = 22;
+int minutes = 15;
+int seconds = 30;
+int nanos = 0;
+ZonedDateTime zdt = ZonedDateTime.of(
+        year, month, day, hour, minutes, seconds, nanos, ZoneId.of('Z'));
+----
++
+* extract pieces from a ZonedDateTime
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 100, ZoneId.of(tz));
+int year = zdt.getYear();
+int month = zdt.getMonthValue();
+int day = zdt.getDayOfMonth();
+int hour = zdt.getHour();
+int minutes = zdt.getMinute();
+int seconds = zdt.getSecond();
+int nanos = zdt.getNano();
+----
+
+==== Datetime Modification
+
+Use either a long or a ZonedDateTime to do datetime modification such as adding
+several seconds to a datetime or subtracting several days from a datetime. Use
+standard <<painless-operators-numeric, numeric operators>> to modify a long
+representation of a datetime. Use ZonedDateTime
+<<painless-api-reference-shared-ZonedDateTime, methods>> to modify a
+ZonedDateTime representation of a datetime. Note most modification methods for
+a ZonedDateTime return a new instance for assignment or immediate use.
+
+===== Datetime Modification Examples
+
+* Subtract three seconds from a long (milliseconds)
```

Comment:
The issue I still have with the "long" examples are calling them longs. As defined in your descriptions above, these are datetimes that are epoch times, which happen to be stored in longs. In other examples you don't use "ZonedDateTime" in the description (which I don't think you should). Here I might say "Subtract three seconds from an epoch datetime". I don't think the milliseconds clarification is necessary then.

Reply:
That's a really good point.  I'll mull over the murkiness of mixing types versus generic datetime.

Reply:
Changed, but I want to follow up with a PR that better addresses this.